File: src/gui/audio_player.py
from tkinter import Misc, Frame, Menu, Label, Button, Listbox, Scrollbar, IntVar, StringVar
from tkinter.ttk import Progressbar, Scale
from tkinter.messagebox import askokcancel, showinfo, showerror
from tkinter.filedialog import askopenfilenames


EXTENSION_LIST = [("Все поддерживаемые форматы", ".wav .flac .m4a .mp3 .ogg"),
                  ("Microsoft Wave", ".wav"),
                  ("Free Lossless Audio Codec", ".flac"),
                  ("Apple Lossless", ".m4a"),
                  ("MPEG Layer 3", ".mp3"),
                  ("Ogg Vorbis Audio", ".ogg")]
TITLE_NAME = "Rust | Python | Audio Player"


class AudioPlayer(Frame):
    def __init__(self, master: Misc = None,):
        super().__init__(master)

        self._volume = IntVar(value=80)
        self._song_pos = IntVar(value=50)
        self._status = StringVar(value="Остановлено")

        self._playlist = list[str]()
        self._playlist_variable = StringVar(value=self._playlist)

        # MASTER
        self.master.title(TITLE_NAME)
        self.master.state("zoomed")
        self.master.protocol("WM_DELETE_WINDOW", self._close)

        # MAIN MENU
        menu = Menu(master)
        self.master.config(menu=menu)

        menu.add_command(label="Добавить", command=self._add)
        menu.add_command(label="Справка", command=self._help)

        # BASE FRAMES
        toolbar = Frame(self.master, relief="raised", borderwidth=1)
        playlist = Frame(self.master)
        statusbar = Frame(self.master, relief="raised", borderwidth=1)

        toolbar.pack(side="top", padx=2, pady=2, fill="x")
        playlist.pack(side="top", fill="both", expand=True)
        statusbar.pack(side="bottom", fill="x")

        # TOOLBAR
        play_button = Button(toolbar, text="PLAY", command=self._play)
        stop_button = Button(toolbar, text="STOP", command=self._stop)
        prev_button = Button(toolbar, text="PREV")
        next_button = Button(toolbar, text="NEXT")

        volume_scale = Scale(toolbar, orient="horizontal", length=100, from_=0, to=100, variable=self._volume,
                             command=lambda e: self._volume.set(value=round(float(e))))

        seek_progressbar = Progressbar(toolbar, mode="determinate", length=300, variable=self._song_pos, maximum=100)

        play_button.pack(side="left", padx=2, pady=2)
        stop_button.pack(side="left", padx=2, pady=2)
        prev_button.pack(side="left", padx=2, pady=2)
        next_button.pack(side="left", padx=2, pady=2)
        volume_scale.pack(side="left", padx=2, pady=2)
        seek_progressbar.pack(side="right", padx=2, pady=2)

        # PLAYLIST
        self._playlist_box = Listbox(playlist, listvariable=self._playlist_variable, selectmode="extended")
        playlist_scroll = Scrollbar(playlist, orient="vertical")

        self._playlist_box.pack(side="left", fill="both", expand=True)
        playlist_scroll.pack(side="right", fill="both")

        self._playlist_box.config(yscrollcommand=playlist_scroll.set)
        playlist_scroll.config(command=self._playlist_box.yview)

        # STATUSBAR
        statusbar_label = Label(statusbar, textvariable=self._status)
        statusbar_label.pack(side="left")

    def _close(self):
        if askokcancel("Выход", "Вы действительно хотите выйти?"):
            self.master.destroy()

    def _add(self):
        filepath = askopenfilenames(title="Выберите один или несколько файлов", filetypes=EXTENSION_LIST)

        if filepath != "":
            for file in filepath:
                self._playlist.append(file)
            self._playlist_variable.set(self._playlist)

    @staticmethod
    def _help():
        showinfo(title="Внимание", message="Спасибо за внимание")

    def _play(self):
        selected_indices = self._playlist_box.curselection()

        if len(selected_indices) > 1:
            showerror(title="Ошибка воспроизведения", message="Невозможно воспроизведение нескольких файлов")
            return

        if len(selected_indices) < 1:
            showerror(title="Ошибка воспроизведения", message="Не выбран объект для воспроизведения")
            return

        self.master.title(self._playlist[selected_indices[0]])
        self._status.set("Играет")

    def _stop(self):
        self.master.title(TITLE_NAME)
        self._status.set("Остановлено")


if __name__ == "__main__":
    app = AudioPlayer()
    app.mainloop()

File: src/gui/audio_player.rs
use gtk::prelude::*;
use gtk::{Align, Application, ApplicationWindow, Box, Button, CenterBox, Label, ListBox, Orientation,
          PolicyType, ProgressBar, ScrolledWindow, VolumeButton};
use gtk::gio::{MenuItem, Menu};


const TITLE_NAME: &str = "Rust | Audio Player";
const STATUS_LIST: &[&str] = &["Is playing", "Is stopped"];


pub(crate) fn build_ui(app: &Application)
{
    // CONTROL BOX
    let control_box = Box::new(Orientation::Horizontal, 10);
    control_box.set_margin_start(6);
    control_box.set_margin_end(6);
    control_box.set_margin_top(6);
    control_box.set_margin_bottom(6);

    let play_button = Button::with_label("PLAY");
    play_button.set_valign(Align::Center);
    play_button.set_halign(Align::Center);

    play_button.connect_clicked(|button| {
        button.set_label("XDDD");
    });

    let stop_button = Button::with_label("STOP");
    stop_button.set_valign(Align::Center);
    stop_button.set_halign(Align::Center);

    stop_button.connect_clicked(|button| {
        button.set_label("WWWW");
    });

    let prev_button = Button::with_label("PREV");
    prev_button.set_valign(Align::Center);
    prev_button.set_halign(Align::Center);

    let next_button = Button::with_label("NEXT");
    next_button.set_valign(Align::Center);
    next_button.set_halign(Align::Center);

    control_box.append(&play_button);
    control_box.append(&stop_button);
    control_box.append(&prev_button);
    control_box.append(&next_button);

    // SONG PROGRESSBAR
    let song_progress = ProgressBar::new();
    song_progress.set_margin_start(6);
    song_progress.set_margin_end(6);
    song_progress.set_margin_top(6);
    song_progress.set_margin_bottom(6);
    song_progress.set_valign(Align::Center);
    song_progress.set_halign(Align::Fill);
    song_progress.set_hexpand(true);

    // VOLUME BUTTON
    let volume_button = VolumeButton::new();
    volume_button.set_margin_start(6);
    volume_button.set_margin_end(6);
    volume_button.set_margin_top(6);
    volume_button.set_margin_bottom(6);
    volume_button.set_valign(Align::Center);
    volume_button.set_halign(Align::Center);
    volume_button.set_value(0.8);

    // TOOLBAR BOX
    let toolbar = CenterBox::new();
    toolbar.set_start_widget(Some(&control_box));
    toolbar.set_center_widget(Some(&song_progress));
    toolbar.set_end_widget(Some(&volume_button));
    toolbar.set_halign(Align::Fill);
    toolbar.set_hexpand(true);

    // PLAYLIST WINDOW
    let playlist = ScrolledWindow::new();
    playlist.set_hexpand(true);
    playlist.set_vexpand(true);
    playlist.set_hscrollbar_policy(PolicyType::Never);
    playlist.set_vscrollbar_policy(PolicyType::Always);

    let playlist_box = ListBox::new();
    playlist.set_child(Some(&playlist_box));

    // STATUSBAR BOX
    let statusbar = Box::new(Orientation::Horizontal, 0);
    statusbar.set_margin_start(6);
    statusbar.set_margin_end(6);
    statusbar.set_margin_top(6);
    statusbar.set_margin_bottom(6);

    let status = Label::with_mnemonic(STATUS_LIST[1]);
    statusbar.append(&status);

    // MENU BAR
    let add = MenuItem::new(Some("Add"), None);
    let help = MenuItem::new(Some("Help"), None);

    let menu = Menu::new();
    menu.append_item(&add);
    menu.append_item(&help);
    app.set_menubar(Some(&menu));

    // WINDOW
    let window = Box::new(Orientation::Vertical, 0);
    window.append(&toolbar);
    window.append(&playlist);
    window.append(&statusbar);

    // APP WINDOW
    let app_window = ApplicationWindow::builder()
        .application(app)
        .title(TITLE_NAME)
        .child(&window)
        .maximized(true)
        .show_menubar(true)
        .build();
    app_window.present();
}